File: infrastructure/observability/tracing/tracing.py
import os
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor

def setup_tracing(app=None):
    """
    Sets up OpenTelemetry Tracer Provider and registers OTLP gRPC exporters.
    Optionally auto-instruments FastAPI app if provided.
    """
    service_name = os.getenv("SERVICE_NAME", "python-service")
    environment = os.getenv("ENV", "production")

    # Define resource metadata
    resource = Resource.create(attributes={
        "service.name": service_name,
        "deployment.environment": environment
    })

    # Set globally scoped Tracer Provider
    provider = TracerProvider(resource=resource)
    trace.set_tracer_provider(provider)

    # Set gRPC trace exporter to OTEL Collector
    collector_endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "otel-collector.monitoring.svc.cluster.local:4317")
    otlp_exporter = OTLPSpanExporter(endpoint=collector_endpoint, insecure=True)

    # Wrap spans in batches to optimize network traffic overhead
    span_processor = BatchSpanProcessor(otlp_exporter)
    provider.add_span_processor(span_processor)

    # Auto-instrument common HTTP request libraries
    RequestsInstrumentor().instrument()

    # If a FastAPI application is specified, apply auto-instrumentation
    if app:
        FastAPIInstrumentor.instrument_app(app)

    return trace.get_tracer(service_name)

File: infrastructure/observability/tracing/tracing.js
const opentelemetry = require('@opentelemetry/sdk-node');
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-grpc');
const { Resource } = require('@opentelemetry/resources');
const { SemanticResourceAttributes } = require('@opentelemetry/semantic-conventions');

// Configure trace exporter pointing to EKS OTEL Collector agent
const traceExporter = new OTLPTraceExporter({
  url: process.env.OTEL_EXPORTER_OTLP_ENDPOINT || 'http://otel-collector.monitoring.svc.cluster.local:4317',
});

// Configure base Node SDK tracing provider
const sdk = new opentelemetry.NodeSDK({
  resource: new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: process.env.SERVICE_NAME || 'node-service',
    [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]: process.env.NODE_ENV || 'production',
  }),
  traceExporter,
  instrumentations: [
    getNodeAutoInstrumentations({
      // Disable noisy core instrumentations if needed
      '@opentelemetry/instrumentation-fs': { enabled: false },
    }),
  ],
});

// Start SDK runtime
sdk.start();

// Handle graceful termination on system shutdown
process.on('SIGTERM', () => {
  sdk.shutdown()
    .then(() => console.log('Tracing SDK terminated successfully.'))
    .catch((err) => console.error('Error terminating Tracing SDK:', err))
    .finally(() => process.exit(0));
});

module.exports = sdk;
